src/analysis/inner_product_space/pi_L2.lean

```diff
@@ -150,12 +152,115 @@ begin
+  (euclidean_space.single i a) j = ite (j = i) a 0 :=
+by { rw [euclidean_space.single, ← pi.single_apply i a j] }
+
+lemma euclidean_space.inner_single_left [decidable_eq ι] (i : ι) (a : 𝕜) (v : euclidean_space 𝕜 ι) :
+  ⟪euclidean_space.single i (a : 𝕜), v⟫ = conj a * (v i) :=
+by simp [apply_ite conj]
+
+lemma euclidean_space.inner_single_right [decidable_eq ι] (i : ι) (a : 𝕜)
+  (v : euclidean_space 𝕜 ι) :
+  ⟪v, euclidean_space.single i (a : 𝕜)⟫ =  a * conj (v i) :=
+by simp [apply_ite conj, mul_comm]
+
+variables (ι 𝕜 E)
+
+/-- An orthonormal basis on E is an identification of `E` with its dimensional-matching
 `euclidean_space 𝕜 ι`. -/
-def basis.isometry_euclidean_of_orthonormal
-  (v : basis ι 𝕜 E) (hv : orthonormal 𝕜 v) :
-  E ≃ₗᵢ[𝕜] euclidean_space 𝕜 ι :=
-v.equiv_fun.isometry_of_inner
+structure orthonormal_basis := of_repr :: (repr : E ≃ₗᵢ[𝕜] euclidean_space 𝕜 ι)
+
+variables {ι 𝕜 E}
+
+namespace orthonormal_basis
+
+instance : inhabited (orthonormal_basis ι 𝕜 (euclidean_space 𝕜 ι)) :=
+⟨of_repr (linear_isometry_equiv.refl 𝕜 (euclidean_space 𝕜 ι))⟩
+
+/-- `b i` is the `i`th basis vector. -/
+instance : has_coe_to_fun (orthonormal_basis ι 𝕜 E) (λ _, ι → E) :=
+{ coe := λ b i, by classical; exact b.repr.symm (euclidean_space.single i (1 : 𝕜)) }
+
+@[simp] protected lemma repr_symm_single [decidable_eq ι] (b : orthonormal_basis ι 𝕜 E) (i : ι) :
+  b.repr.symm (euclidean_space.single i (1:𝕜)) = b i :=
+by { classical, congr, simp, }
+
+@[simp] protected lemma repr_self [decidable_eq ι] (b : orthonormal_basis ι 𝕜 E) (i : ι) :
+  b.repr (b i) = euclidean_space.single i (1:𝕜) :=
+begin
+  classical,
+  rw [← b.repr_symm_single i, linear_isometry_equiv.apply_symm_apply],
+  congr,
+  simp,
+end
+
+protected lemma repr_apply_apply (b : orthonormal_basis ι 𝕜 E) (v : E) (i : ι) :
+  b.repr v i = ⟪b i, v⟫ :=
+begin
+  classical,
+  rw [← b.repr.inner_map_map (b i) v, b.repr_self i, euclidean_space.inner_single_left],
+  simp only [one_mul, eq_self_iff_true, map_one],
+end
+
+@[simp]
+protected lemma orthonormal (b : orthonormal_basis ι 𝕜 E) : orthonormal 𝕜 b :=
+begin
+  classical,
+  rw orthonormal_iff_ite,
+  intros i j,
+  rw [← b.repr.inner_map_map (b i) (b j), b.repr_self i, b.repr_self j],
+  rw euclidean_space.inner_single_left,
+  rw euclidean_space.single_apply,
+  simp only [mul_boole, map_one],
+end
+
+/- The `basis ι 𝕜 E` underlying the `orthonormal_basis` --/
+protected def to_basis (b : orthonormal_basis ι 𝕜 E) : basis ι 𝕜 E :=
+  basis.of_equiv_fun b.repr.to_linear_equiv
+
+@[simp] protected lemma coe_to_basis (b : orthonormal_basis ι 𝕜 E) :
+  (⇑b.to_basis : ι → E) = ⇑b :=
+begin
+  change ⇑(basis.of_equiv_fun b.repr.to_linear_equiv) = b,
+  ext j,
+  rw basis.coe_of_equiv_fun,
+  simp only [orthonormal_basis.repr_symm_single],
+  congr,
+end
+
+@[simp] protected lemma to_basis_orthonormal (b : orthonormal_basis ι 𝕜 E) :
+  orthonormal 𝕜 (b.to_basis) :=
```

Comment:
I'd honestly just delete this lemma, since you can always get it by simp and I'm not sure `⇑b.to_basis` would turn up in the wild.